File: Deepak Urs/Arrays/nextPermutation.java
class Solution {
    private void reverse(int st, int en, int [] nums) {
        while(st < en) {
            int temp = nums[st];
            nums[st] = nums[en];
            nums[en] = temp;
            st++;
            en--;
        }
    }

    private void swap(int st, int en, int [] nums) {
        int temp = nums[st];
        nums[st] = nums[en];
        nums[en] = temp;
    }

    public void nextPermutation(int[] nums) {
       int i = nums.length - 2;

       while(i >=0 && nums[i] >= nums[i+1]){
            i--;
       } 

       if(i == -1) {
        reverse(0,nums.length-1, nums);
       }
       else {
        int j = nums.length-1;

        while(j > i && nums[j] <= nums[i]) {
            j--;
        }

        swap(i, j, nums);
        reverse(i+1, nums.length-1, nums);
       }
    }
}

File: Abubakr Khanooni/Arrays/nextPermutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int goal_index = -1;
        for(int i = n - 1; i > 0; i--) {
            if(nums[i] > nums[i - 1]) {
                goal_index = i - 1;
                break;
            }
        }
        if(goal_index != -1) {
            int swap_index = goal_index;
            for(int j = n - 1; j > goal_index; j--) {
                if(nums[j] > nums[goal_index]) {
                    swap_index = j;
                    break;
                }
            }
            swap(nums[goal_index], nums[swap_index]);
        }
        reverse(nums.begin() + goal_index + 1, nums.end());
    }
};